# **2025 March Madness Bracket Predictor**

## 1. Data Collection

### 1.1 Importing Current KenPom statistics with Matchup Data



In [46]:
import pandas as pd

kenPom = pd.read_csv('kenpom-ncaa-2025.csv')
kenPom.head()

# Drop columns that are not needed  
kenPom.dropna()


Unnamed: 0     Rk                    Team  Conf  NetRtg   ORtg   DRtg  \
0             0    1.0                    Duke   ACC   39.40  129.1   89.7   
1             1    2.0                  Auburn   SEC   37.67  130.7   93.0   
2             2    3.0                 Houston   B12   35.14  124.4   89.2   
3             3    4.0                 Florida   SEC   33.97  126.8   92.9   
4             4    5.0               Tennessee   SEC   31.40  118.5   87.1   
..          ...    ...                     ...   ...     ...    ...    ...   
341         341  358.0                Canisius  MAAC  -22.12  100.8  122.9   
342         342  359.0             Alabama A&M  SWAC  -22.34   93.2  115.5   
343         343  360.0  Maryland Eastern Shore  MEAC  -23.37   95.2  118.5   
344         344  363.0     Arkansas Pine Bluff  SWAC  -26.82   95.1  122.0   
345         345  364.0  Mississippi Valley St.  SWAC  -44.57   82.1  126.6   

     AdjT   Luck  SS_NetRtg  SS_ORtg  SS_DRtg  NCSOS_NetRtg  
0    66.3 -0.043       9.18    112.8    103.7          9.96  
1    68.6  0.063      18.31    117.2     98.9         11.00  
2    61.3 -0.001      13.20    113.4    100.2          3.25  
3    69.5 -0.005      12.32    113.4    101.1         -1.18  
4    64.0  0.030      14.83    116.7    101.9         -0.36  
..    ...    ...        ...      ...      ...           ...  
341  66.1 -0.080      -4.10    103.8    107.9          4.02  
342  72.0 -0.019      -9.37    102.3    111.7          2.54  
343  66.2 -0.105      -2.28    106.4    108.7         10.24  
344  72.2 -0.058      -7.40    102.3    109.7          7.54  
345  64.7  0.001      -1.60    105.4    107.0         19.28  

[346 rows x 13 columns]

In [47]:
#Describe the data
kenPom.info()
kenPom.describe()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 346 entries, 0 to 345
Data columns (total 13 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Unnamed: 0    346 non-null    int64  
 1   Rk            346 non-null    float64
 2   Team          346 non-null    object 
 3   Conf          346 non-null    object 
 4   NetRtg        346 non-null    float64
 5   ORtg          346 non-null    float64
 6   DRtg          346 non-null    float64
 7   AdjT          346 non-null    float64
 8   Luck          346 non-null    float64
 9   SS_NetRtg     346 non-null    float64
 10  SS_ORtg       346 non-null    float64
 11  SS_DRtg       346 non-null    float64
 12  NCSOS_NetRtg  346 non-null    float64
dtypes: float64(10), int64(1), object(2)
memory usage: 35.3+ KB


Unnamed: 0          Rk      NetRtg        ORtg        DRtg        AdjT  \
count   346.00000  346.000000  346.000000  346.000000  346.000000  346.000000   
mean    172.50000  181.511561    0.161214  107.439595  107.279191   67.553468   
std     100.02583  105.228178   12.873322    7.542876    6.776453    2.506722   
min       0.00000    1.000000  -44.570000   82.100000   87.100000   59.400000   
25%      86.25000   91.250000   -9.192500  101.800000  103.000000   65.800000   
50%     172.50000  181.500000   -1.505000  106.900000  107.550000   67.600000   
75%     258.75000  271.750000    7.755000  112.275000  112.000000   69.200000   
max     345.00000  364.000000   39.400000  130.700000  126.600000   75.000000   

             Luck   SS_NetRtg     SS_ORtg     SS_DRtg  NCSOS_NetRtg  
count  346.000000  346.000000  346.000000  346.000000    346.000000  
mean     0.001884    1.067457  107.853179  106.787283      0.725000  
std      0.056944    6.691981    3.637678    3.264497      5.194638  
min     -0.176000  -10.230000  100.700000   98.600000    -11.140000  
25%     -0.034000   -3.397500  105.500000  104.600000     -2.597500  
50%     -0.001500   -1.075000  107.100000  107.800000      0.415000  
75%      0.038500    5.417500  109.875000  109.375000      3.297500  
max      0.190000   18.310000  117.700000  111.900000     20.440000

In [48]:
# Create all possible matchups
from itertools import combinations

matchups = list(combinations(kenPom['Team'], 2))

matchup_df = pd.DataFrame(matchups, columns=['Team1', 'Team2'])    

matchup_df.head()



Team1      Team2
0  Duke     Auburn
1  Duke    Houston
2  Duke    Florida
3  Duke  Tennessee
4  Duke    Alabama

In [49]:
#Merge Team 1 and Team 2 KenPom stats into the matchup_df
df_team1 = kenPom.rename(columns={col: f"Team_1_{col}" for col in kenPom.columns if col not in ["Team"]})
df_team2 = kenPom.rename(columns={col: f"Team_2_{col}" for col in kenPom.columns if col not in ["Team"]})

df_team1 = df_team1.rename(columns={"Team": "Team1"})
df_team2 = df_team2.rename(columns={"Team": "Team2"})



#coloumn name check
"""
print("Matchup DataFrame Columns:", matchup_df.columns)
print("KenPom DataFrame Columns (df):", kenPom.columns)
print("df_team1 Columns:", df_team1.columns)
print("df_team2 Columns:", df_team2.columns)
"""
matchup_df = matchup_df.merge(df_team1, on='Team1', how = 'left')
matchup_df = matchup_df.merge(df_team2, on='Team2', how = 'left')   

matchup_df.head()
matchup_df.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 59685 entries, 0 to 59684
Data columns (total 26 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   Team1                59685 non-null  object 
 1   Team2                59685 non-null  object 
 2   Team_1_Unnamed: 0    59685 non-null  int64  
 3   Team_1_Rk            59685 non-null  float64
 4   Team_1_Conf          59685 non-null  object 
 5   Team_1_NetRtg        59685 non-null  float64
 6   Team_1_ORtg          59685 non-null  float64
 7   Team_1_DRtg          59685 non-null  float64
 8   Team_1_AdjT          59685 non-null  float64
 9   Team_1_Luck          59685 non-null  float64
 10  Team_1_SS_NetRtg     59685 non-null  float64
 11  Team_1_SS_ORtg       59685 non-null  float64
 12  Team_1_SS_DRtg       59685 non-null  float64
 13  Team_1_NCSOS_NetRtg  59685 non-null  float64
 14  Team_2_Unnamed: 0    59685 non-null  int64  
 15  Team_2_Rk            59685 non-null 

In [50]:
#computing feature differences between two teams in each possible matchup:
# Compute feature differences between Team_1 and Team_2
matchup_df["ORtg_Diff"] = matchup_df["Team_1_ORtg"] - matchup_df["Team_2_ORtg"]
matchup_df["DRtg_Diff"] = matchup_df["Team_1_DRtg"] - matchup_df["Team_2_DRtg"]
matchup_df["AdjT_Diff"] = matchup_df["Team_1_AdjT"] - matchup_df["Team_2_AdjT"]
matchup_df["Luck_Diff"] = matchup_df["Team_1_Luck"] - matchup_df["Team_2_Luck"]
matchup_df["SS_NetRtg_Diff"] = matchup_df["Team_1_SS_NetRtg"] - matchup_df["Team_2_SS_NetRtg"]
matchup_df.head()

Team1      Team2  Team_1_Unnamed: 0  Team_1_Rk Team_1_Conf  Team_1_NetRtg  \
0  Duke     Auburn                  0        1.0         ACC           39.4   
1  Duke    Houston                  0        1.0         ACC           39.4   
2  Duke    Florida                  0        1.0         ACC           39.4   
3  Duke  Tennessee                  0        1.0         ACC           39.4   
4  Duke    Alabama                  0        1.0         ACC           39.4   

   Team_1_ORtg  Team_1_DRtg  Team_1_AdjT  Team_1_Luck  ...  Team_2_Luck  \
0        129.1         89.7         66.3       -0.043  ...        0.063   
1        129.1         89.7         66.3       -0.043  ...       -0.001   
2        129.1         89.7         66.3       -0.043  ...       -0.005   
3        129.1         89.7         66.3       -0.043  ...        0.030   
4        129.1         89.7         66.3       -0.043  ...        0.029   

   Team_2_SS_NetRtg  Team_2_SS_ORtg  Team_2_SS_DRtg  Team_2_NCSOS_NetRtg  \
0             18.31           117.2            98.9                11.00   
1             13.20           113.4           100.2                 3.25   
2             12.32           113.4           101.1                -1.18   
3             14.83           116.7           101.9                -0.36   
4             18.25           117.7            99.5                12.59   

   ORtg_Diff DRtg_Diff  AdjT_Diff  Luck_Diff  SS_NetRtg_Diff  
0       -1.6      -3.3       -2.3     -0.106           -9.13  
1        4.7       0.5        5.0     -0.042           -4.02  
2        2.3      -3.2       -3.2     -0.038           -3.14  
3       10.6       2.6        2.3     -0.073           -5.65  
4        1.8      -8.1       -8.7     -0.072           -9.07  

[5 rows x 31 columns]

## 1.2 Importing Historical KenPom Stats with Tournament matchup and result data

In [51]:
histKenPom = pd.read_csv('mmHistData/KenPom Barttorvik.csv')
histKenPom.head()

YEAR  CONF  CONF ID  QUAD NO  QUAD ID  TEAM NO  TEAM ID         TEAM  SEED  \
0  2025   MAC       17       66        2     1147        2        Akron    13   
1  2025   SEC       28       66        2     1146        3      Alabama     2   
2  2025  SWAC       31       68        4     1145        4  Alabama St.    16   
3  2025   Pat       25       66        2     1144        6     American    16   
4  2025   B12        7       66        2     1143        8      Arizona     4   

   ROUND  ...  BADJT RANK  AVG HGT RANK  EFF HGT RANK  EXP RANK  TALENT RANK  \
0      0  ...          19           361           364        73          149   
1      0  ...           1            57             4       261           32   
2      0  ...         154           285           239        83          251   
3      0  ...         349           303           248       181          364   
4      0  ...          52            74           129       225           16   

   FT% RANK  OP FT% RANK  PPPO RANK  PPPD RANK  ELITE SOS RANK  
0        86          239         25        145             305  
1       211           92          9        174               1  
2       270           61        227        147             253  
3        58           34        218        163             329  
4        18          143         25         87               3  

[5 rows x 103 columns]

In [52]:
#show columns
histKenPom.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1147 entries, 0 to 1146
Columns: 103 entries, YEAR to ELITE SOS RANK
dtypes: float64(46), int64(55), object(2)
memory usage: 923.1+ KB


In [53]:
#Keep relevant coloumns and clean histKenPom

columns_to_keep = ['YEAR', 'TEAM', 'SEED', 'KADJ O', 'KADJ D', 'K TEMPO']

histKenpom_df= histKenPom[columns_to_keep]

#rename columns for merging
histKenpom_df = histKenpom_df.rename(columns={"TEAM": "Team"})
histKenpom_df = histKenpom_df.rename(columns={"YEAR": "Year"})
histKenpom_df = histKenpom_df.rename(columns={"SEED": "Seed"})
histKenpom_df = histKenpom_df.rename(columns={"KADJ O": "AdjO"})
histKenpom_df = histKenpom_df.rename(columns={"KADJ D": "AdjD"})
histKenpom_df = histKenpom_df.rename(columns={"K TEMPO": "AdjT"})

histKenpom_df.tail()

Year              Team  Seed      AdjO     AdjD     AdjT
1142  2008     West Virginia     7  113.5700  93.4302  66.2223
1143  2008  Western Kentucky    12  111.0760  97.0110  68.7887
1144  2008          Winthrop    13   99.6932  92.3864  63.7666
1145  2008         Wisconsin     3  113.6190  86.2081  62.1920
1146  2008            Xavier     3  116.4130  94.0680  65.1599

In [54]:
histMatchups = pd.read_csv('mmHistData/Tournament Matchups.csv')
histMatchups.head()

YEAR  BY YEAR NO  BY ROUND NO  TEAM NO              TEAM  SEED  ROUND  \
0  2024        2014         2094     1067       Connecticut     1      1   
1  2024        2013         2093     1026           Stetson    16     64   
2  2024        2012         2092     1060  Florida Atlantic     8     64   
3  2024        2011         2091     1036      Northwestern     9     32   
4  2024        2010         2090     1029     San Diego St.     5     16   

   CURRENT ROUND  SCORE  
0             64     91  
1             64     52  
2             64     65  
3             64     77  
4             64     69

In [55]:
histMatchups.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2014 entries, 0 to 2013
Data columns (total 9 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   YEAR           2014 non-null   int64 
 1   BY YEAR NO     2014 non-null   int64 
 2   BY ROUND NO    2014 non-null   int64 
 3   TEAM NO        2014 non-null   int64 
 4   TEAM           2014 non-null   object
 5   SEED           2014 non-null   int64 
 6   ROUND          2014 non-null   int64 
 7   CURRENT ROUND  2014 non-null   int64 
 8   SCORE          2014 non-null   int64 
dtypes: int64(8), object(1)
memory usage: 141.7+ KB


In [56]:
#filter out irrelevant columns
columns_to_keep = ['YEAR', 'TEAM', 'SCORE', 'ROUND']

histMatchups_df = histMatchups[columns_to_keep]

histMatchups_df.head()

YEAR              TEAM  SCORE  ROUND
0  2024       Connecticut     91      1
1  2024           Stetson     52     64
2  2024  Florida Atlantic     65     64
3  2024      Northwestern     77     32
4  2024     San Diego St.     69     16

In [57]:
matchups = []

for i in range(0, len(histMatchups_df), 2):  # Step by 2 to group consecutive pairs
    if i + 1 < len(histMatchups_df):  # Ensure we don't go out of bounds
        matchups.append({
            'YEAR': histMatchups_df.loc[i, 'YEAR'],
            'Team_1': histMatchups_df.loc[i, 'TEAM'],
            'Score_1': histMatchups_df.loc[i, 'SCORE'],
            'Team_2': histMatchups_df.loc[i + 1, 'TEAM'],
            'Score_2': histMatchups_df.loc[i + 1, 'SCORE'],
            'Winner': int(histMatchups_df.loc[i, 'SCORE'] > histMatchups_df.loc[i + 1, 'SCORE'])  # 1 if Team_1 wins, else 0
        })

# Convert the list into a DataFrame
matchups_df = pd.DataFrame(matchups)

matchups_df.head()



YEAR            Team_1  Score_1        Team_2  Score_2  Winner
0  2024       Connecticut       91       Stetson       52       1
1  2024  Florida Atlantic       65  Northwestern       77       0
2  2024     San Diego St.       69           UAB       65       1
3  2024            Auburn       76          Yale       78       0
4  2024               BYU       67      Duquesne       71       0

In [58]:
#Remove 2025 entries from histKenpom_df
histKenpom_df = histKenpom_df[histKenpom_df['Year'] != 2025]
histKenpom_df.head()

Year     Team  Seed     AdjO      AdjD     AdjT
68  2024    Akron    14  107.009  102.0720  66.7747
69  2024  Alabama     4  125.601  103.1180  74.1625
70  2024  Arizona     2  121.125   94.5013  73.3760
71  2024   Auburn     4  120.579   91.6765  70.9629
72  2024   Baylor     3  122.490  100.1850  66.8428

In [59]:
#Merge histKenpom_df with histMatchups_df

#Team 1
merged_df = matchups_df.merge(histKenpom_df, left_on=['YEAR', 'Team_1'], right_on=['Year', 'Team'], how='left')

#Team 2
merged_df = merged_df.merge(histKenpom_df, left_on=['Year', 'Team_2'], right_on=['Year', 'Team'], how='left')

merged_df.head()


YEAR            Team_1  Score_1        Team_2  Score_2  Winner    Year  \
0  2024       Connecticut       91       Stetson       52       1  2024.0   
1  2024  Florida Atlantic       65  Northwestern       77       0  2024.0   
2  2024     San Diego St.       69           UAB       65       1  2024.0   
3  2024            Auburn       76          Yale       78       0  2024.0   
4  2024               BYU       67      Duquesne       71       0  2024.0   

             Team_x  Seed_x   AdjO_x    AdjD_x   AdjT_x        Team_y  Seed_y  \
0       Connecticut     1.0  126.608   94.3949  66.2316       Stetson    16.0   
1  Florida Atlantic     8.0  118.961  102.7250  69.0429  Northwestern     9.0   
2     San Diego St.     5.0  113.548   93.8492  67.0337           UAB    12.0   
3            Auburn     4.0  120.579   91.6765  70.9629          Yale    13.0   
4               BYU     6.0  120.486   99.4765  70.0246      Duquesne    11.0   

    AdjO_y    AdjD_y   AdjT_y  
0  110.262  114.3660  66.6291  
1  116.657  100.7520  64.6113  
2  113.680  107.3980  68.7165  
3  110.825  101.3520  65.6179  
4  106.921   97.5497  66.8316

In [60]:
# Rename columns for clarity
merged_df.rename(columns={
    'Team_x': 'Team_1',
    'Seed_x': 'Seed_1',
    'AdjO_x': 'AdjO_Team1',
    'AdjD_x': 'AdjD_Team1',
    'AdjT_x': 'AdjT_Team1',
    'Team_y': 'Team_2',
    'Seed_y': 'Seed_2',
    'AdjO_y': 'AdjO_Team2',
    'AdjD_y': 'AdjD_Team2',
    'AdjT_y': 'AdjT_Team2'
}, inplace=True)

merged_df.head()

YEAR            Team_1  Score_1        Team_2  Score_2  Winner    Year  \
0  2024       Connecticut       91       Stetson       52       1  2024.0   
1  2024  Florida Atlantic       65  Northwestern       77       0  2024.0   
2  2024     San Diego St.       69           UAB       65       1  2024.0   
3  2024            Auburn       76          Yale       78       0  2024.0   
4  2024               BYU       67      Duquesne       71       0  2024.0   

             Team_1  Seed_1  AdjO_Team1  AdjD_Team1  AdjT_Team1        Team_2  \
0       Connecticut     1.0     126.608     94.3949     66.2316       Stetson   
1  Florida Atlantic     8.0     118.961    102.7250     69.0429  Northwestern   
2     San Diego St.     5.0     113.548     93.8492     67.0337           UAB   
3            Auburn     4.0     120.579     91.6765     70.9629          Yale   
4               BYU     6.0     120.486     99.4765     70.0246      Duquesne   

   Seed_2  AdjO_Team2  AdjD_Team2  AdjT_Team2  
0    16.0     110.262    114.3660     66.6291  
1     9.0     116.657    100.7520     64.6113  
2    12.0     113.680    107.3980     68.7165  
3    13.0     110.825    101.3520     65.6179  
4    11.0     106.921     97.5497     66.8316

In [61]:
#drop unnecessary columns
columns_to_drop = ['YEAR', 'Year', 'Score_1', 'Score_2']
merged_df.drop(columns=columns_to_drop, inplace=True)

merged_df.head()

Team_1        Team_2  Winner            Team_1  Seed_1  \
0       Connecticut       Stetson       1       Connecticut     1.0   
1  Florida Atlantic  Northwestern       0  Florida Atlantic     8.0   
2     San Diego St.           UAB       1     San Diego St.     5.0   
3            Auburn          Yale       0            Auburn     4.0   
4               BYU      Duquesne       0               BYU     6.0   

   AdjO_Team1  AdjD_Team1  AdjT_Team1        Team_2  Seed_2  AdjO_Team2  \
0     126.608     94.3949     66.2316       Stetson    16.0     110.262   
1     118.961    102.7250     69.0429  Northwestern     9.0     116.657   
2     113.548     93.8492     67.0337           UAB    12.0     113.680   
3     120.579     91.6765     70.9629          Yale    13.0     110.825   
4     120.486     99.4765     70.0246      Duquesne    11.0     106.921   

   AdjD_Team2  AdjT_Team2  
0    114.3660     66.6291  
1    100.7520     64.6113  
2    107.3980     68.7165  
3    101.3520     65.6179  
4     97.5497     66.8316

In [62]:
merged_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1007 entries, 0 to 1006
Data columns (total 13 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   Team_1      1007 non-null   object 
 1   Team_2      1007 non-null   object 
 2   Winner      1007 non-null   int64  
 3   Team_1      1002 non-null   object 
 4   Seed_1      1002 non-null   float64
 5   AdjO_Team1  1002 non-null   float64
 6   AdjD_Team1  1002 non-null   float64
 7   AdjT_Team1  1002 non-null   float64
 8   Team_2      998 non-null    object 
 9   Seed_2      998 non-null    float64
 10  AdjO_Team2  998 non-null    float64
 11  AdjD_Team2  998 non-null    float64
 12  AdjT_Team2  998 non-null    float64
dtypes: float64(8), int64(1), object(4)
memory usage: 102.4+ KB


In [63]:
#Calculating difference in historical KenPom stats between Team 1 and Team 2
# Compute stat differences
merged_df['AdjO_Diff'] = merged_df['AdjO_Team1'] - merged_df['AdjO_Team2']
merged_df['AdjD_Diff'] = merged_df['AdjD_Team1'] - merged_df['AdjD_Team2']
merged_df['AdjT_Diff'] = merged_df['AdjT_Team1'] - merged_df['AdjT_Team2']
merged_df['Seed_Diff'] = merged_df['Seed_1'] - merged_df['Seed_2']

# Drop the original stat columns (optional, to keep only the differences)
columns_to_drop = ['AdjO_Team1', 'AdjD_Team1', 'AdjT_Team1', 'AdjO_Team2', 'AdjD_Team2', 'AdjT_Team2']
merged_df.drop(columns=columns_to_drop, inplace=True)

# Inspect final dataset
print(merged_df.head())


             Team_1        Team_2  Winner            Team_1  Seed_1  \
0       Connecticut       Stetson       1       Connecticut     1.0   
1  Florida Atlantic  Northwestern       0  Florida Atlantic     8.0   
2     San Diego St.           UAB       1     San Diego St.     5.0   
3            Auburn          Yale       0            Auburn     4.0   
4               BYU      Duquesne       0               BYU     6.0   

         Team_2  Seed_2  AdjO_Diff  AdjD_Diff  AdjT_Diff  Seed_Diff  
0       Stetson    16.0     16.346   -19.9711    -0.3975      -15.0  
1  Northwestern     9.0      2.304     1.9730     4.4316       -1.0  
2           UAB    12.0     -0.132   -13.5488    -1.6828       -7.0  
3          Yale    13.0      9.754    -9.6755     5.3450       -9.0  
4      Duquesne    11.0     13.565     1.9268     3.1930       -5.0  


In [ ]:
#Drop unnecessary columns
columns_to_drop = ['Seed_1', 'Seed_2']
merged_df.drop(columns=columns_to_drop, inplace=True)


columns_to_drop = ['Team_1', 'Team_2']
merged_df = merged_df.loc[:, ~merged_df.columns.duplicated()]


merged_df.head()

             Team_1        Team_2  Winner  AdjO_Diff  AdjD_Diff  AdjT_Diff  \
0       Connecticut       Stetson       1     16.346   -19.9711    -0.3975   
1  Florida Atlantic  Northwestern       0      2.304     1.9730     4.4316   
2     San Diego St.           UAB       1     -0.132   -13.5488    -1.6828   
3            Auburn          Yale       0      9.754    -9.6755     5.3450   
4               BYU      Duquesne       0     13.565     1.9268     3.1930   

   Seed_Diff  
0      -15.0  
1       -1.0  
2       -7.0  
3       -9.0  
4       -5.0  


Team_1        Team_2  Winner  AdjO_Diff  AdjD_Diff  AdjT_Diff  \
0       Connecticut       Stetson       1     16.346   -19.9711    -0.3975   
1  Florida Atlantic  Northwestern       0      2.304     1.9730     4.4316   
2     San Diego St.           UAB       1     -0.132   -13.5488    -1.6828   
3            Auburn          Yale       0      9.754    -9.6755     5.3450   
4               BYU      Duquesne       0     13.565     1.9268     3.1930   

   Seed_Diff  
0      -15.0  
1       -1.0  
2       -7.0  
3       -9.0  
4       -5.0

In [75]:
#Drop rows with missing values
merged_df = merged_df.dropna()

## 2. Training Different Models

### 2.1 Initializing and splitting Data

In [76]:
from sklearn.model_selection import train_test_split

#Defining features(x) and target variable(y)
X = merged_df[['AdjO_Diff', 'AdjD_Diff', 'AdjT_Diff', 'Seed_Diff']]
y = merged_df['Winner']

#Splitting the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42, stratify=y)

# Print dataset sizes
print(f"Training size: {len(X_train)}, Testing size: {len(X_test)}")

Training size: 798, Testing size: 200


### 2.2 Training and Testing Linear Regression Model

In [ ]:
from sklearn.linear_model import LogisticRegression

model = LogisticRegression()

model.fit(X_train, y_train)

train_accuracy = model.score(X_train, y_train)

print(f"Training Accuracy: {train_accuracy:.4f}")

Training Accuracy: 0.7155


In [78]:
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report

# Make predictions on the test set
y_pred = model.predict(X_test)

# Calculate test accuracy
test_accuracy = accuracy_score(y_test, y_pred)
print(f"Test Accuracy: {test_accuracy:.4f}")

# Print classification report (detailed metrics)
print("\nClassification Report:\n", classification_report(y_test, y_pred))

# Print confusion matrix
print("\nConfusion Matrix:\n", confusion_matrix(y_test, y_pred))


Test Accuracy: 0.6700

Classification Report:
               precision    recall  f1-score   support

           0       0.59      0.42      0.49        76
           1       0.70      0.82      0.76       124

    accuracy                           0.67       200
   macro avg       0.65      0.62      0.62       200
weighted avg       0.66      0.67      0.66       200


Confusion Matrix:
 [[ 32  44]
 [ 22 102]]


### 2.3 Training and Testing Random Forest Model

In [79]:
from sklearn.ensemble import RandomForestClassifier

# Initialize the Random Forest model
rf_model = RandomForestClassifier(random_state=42)

# Train the model on the training data
rf_model.fit(X_train, y_train)

# Print training accuracy
train_accuracy_rf = rf_model.score(X_train, y_train)
print(f"Training Accuracy (Random Forest): {train_accuracy_rf:.4f}")


Training Accuracy (Random Forest): 1.0000


In [80]:
# Make predictions on the test set
y_pred_rf = rf_model.predict(X_test)

# Calculate test accuracy
test_accuracy_rf = accuracy_score(y_test, y_pred_rf)
print(f"Test Accuracy (Random Forest): {test_accuracy_rf:.4f}")

# Print the classification report
print("\nClassification Report (Random Forest):\n", classification_report(y_test, y_pred_rf))

# Print confusion matrix
print("\nConfusion Matrix (Random Forest):\n", confusion_matrix(y_test, y_pred_rf))


Test Accuracy (Random Forest): 0.6400

Classification Report (Random Forest):
               precision    recall  f1-score   support

           0       0.54      0.36      0.43        76
           1       0.67      0.81      0.74       124

    accuracy                           0.64       200
   macro avg       0.61      0.58      0.58       200
weighted avg       0.62      0.64      0.62       200


Confusion Matrix (Random Forest):
 [[ 27  49]
 [ 23 101]]


In [88]:
from sklearn.model_selection import RandomizedSearchCV
import numpy as np

# Define parameter grid for Random Forest
param_dist_rf = {
    'n_estimators': [50, 100, 200],
    'max_depth': [None, 10, 20, 30],
    'min_samples_split': [2, 5, 10],
    'min_samples_leaf': [1, 2, 4],
    'max_features': ['sqrt', 'log2']  # Valid options
}

# Initialize Random Forest model
rf_model = RandomForestClassifier(random_state=42)

# Initialize RandomizedSearchCV
random_search_rf = RandomizedSearchCV(estimator=rf_model, param_distributions=param_dist_rf, n_iter=100, cv=3, n_jobs=-1, verbose=2)

# Fit the model
random_search_rf.fit(X_train, y_train)

# Print the best parameters
print("Best parameters for Random Forest:", random_search_rf.best_params_)

# Get the best model
best_rf_model = random_search_rf.best_estimator_

# Print test accuracy of the best model
test_accuracy_rf_tuned = best_rf_model.score(X_test, y_test)
print(f"Tuned Random Forest Test Accuracy: {test_accuracy_rf_tuned:.4f}")


Fitting 3 folds for each of 100 candidates, totalling 300 fits
Best parameters for Random Forest: {'n_estimators': 50, 'min_samples_split': 5, 'min_samples_leaf': 4, 'max_features': 'log2', 'max_depth': 10}
Tuned Random Forest Test Accuracy: 0.6500


### 2.4 Training an Testing XGBoost Model

In [81]:
import xgboost as xgb

# Initialize XGBoost classifier
xgb_model = xgb.XGBClassifier(random_state=42, use_label_encoder=False, eval_metric='logloss')

# Train the model
xgb_model.fit(X_train, y_train)

# Calculate training accuracy
train_accuracy_xgb = xgb_model.score(X_train, y_train)
print(f"Training Accuracy (XGBoost): {train_accuracy_xgb:.4f}")

Training Accuracy (XGBoost): 1.0000


C:\Users\shash\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\LocalCache\local-packages\Python311\site-packages\xgboost\training.py:183: UserWarning: [21:51:06] WARNING: C:\actions-runner\_work\xgboost\xgboost\src\learner.cc:738: 
Parameters: { "use_label_encoder" } are not used.

  bst.update(dtrain, iteration=i, fobj=obj)


In [82]:
# Make predictions on the test set
y_pred_xgb = xgb_model.predict(X_test)

# Calculate test accuracy
test_accuracy_xgb = accuracy_score(y_test, y_pred_xgb)
print(f"Test Accuracy (XGBoost): {test_accuracy_xgb:.4f}")

# Print classification report
print("\nClassification Report (XGBoost):\n", classification_report(y_test, y_pred_xgb))

# Print confusion matrix
print("\nConfusion Matrix (XGBoost):\n", confusion_matrix(y_test, y_pred_xgb))


Test Accuracy (XGBoost): 0.6450

Classification Report (XGBoost):
               precision    recall  f1-score   support

           0       0.55      0.38      0.45        76
           1       0.68      0.81      0.74       124

    accuracy                           0.65       200
   macro avg       0.61      0.59      0.59       200
weighted avg       0.63      0.65      0.63       200


Confusion Matrix (XGBoost):
 [[ 29  47]
 [ 24 100]]


In [83]:
# Make predictions on the test set
y_pred_xgb = xgb_model.predict(X_test)

# Calculate test accuracy
test_accuracy_xgb = accuracy_score(y_test, y_pred_xgb)
print(f"Test Accuracy (XGBoost): {test_accuracy_xgb:.4f}")

# Print classification report
print("\nClassification Report (XGBoost):\n", classification_report(y_test, y_pred_xgb))

# Print confusion matrix
print("\nConfusion Matrix (XGBoost):\n", confusion_matrix(y_test, y_pred_xgb))


Test Accuracy (XGBoost): 0.6450

Classification Report (XGBoost):
               precision    recall  f1-score   support

           0       0.55      0.38      0.45        76
           1       0.68      0.81      0.74       124

    accuracy                           0.65       200
   macro avg       0.61      0.59      0.59       200
weighted avg       0.63      0.65      0.63       200


Confusion Matrix (XGBoost):
 [[ 29  47]
 [ 24 100]]


In [85]:
from sklearn.model_selection import RandomizedSearchCV
import xgboost as xgb

# Define parameter grid for XGBoost (using ranges)
param_dist_xgb = {
    'n_estimators': [50, 100, 200, 300],
    'max_depth': [3, 6, 9, 12],
    'learning_rate': [0.01, 0.1, 0.2],
    'subsample': [0.7, 0.8, 0.9, 1.0],
    'colsample_bytree': [0.7, 0.8, 0.9],
    'gamma': [0, 0.1, 0.2]
}

# Initialize XGBoost model
xgb_model = xgb.XGBClassifier(random_state=42, use_label_encoder=False, eval_metric='logloss')

# Initialize RandomizedSearchCV
random_search_xgb = RandomizedSearchCV(estimator=xgb_model, param_distributions=param_dist_xgb, n_iter=100, cv=3, n_jobs=-1, verbose=2)

# Fit the model
random_search_xgb.fit(X_train, y_train)

# Print best parameters
print("Best parameters for XGBoost:", random_search_xgb.best_params_)

# Get the best model
best_xgb_model = random_search_xgb.best_estimator_

# Print test accuracy of the best model
test_accuracy_xgb_tuned = best_xgb_model.score(X_test, y_test)
print(f"Tuned XGBoost Test Accuracy: {test_accuracy_xgb_tuned:.4f}")



Fitting 3 folds for each of 100 candidates, totalling 300 fits
Best parameters for XGBoost: {'subsample': 0.9, 'n_estimators': 200, 'max_depth': 3, 'learning_rate': 0.01, 'gamma': 0.1, 'colsample_bytree': 0.9}
Tuned XGBoost Test Accuracy: 0.6950


C:\Users\shash\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\LocalCache\local-packages\Python311\site-packages\xgboost\training.py:183: UserWarning: [22:08:03] WARNING: C:\actions-runner\_work\xgboost\xgboost\src\learner.cc:738: 
Parameters: { "use_label_encoder" } are not used.

  bst.update(dtrain, iteration=i, fobj=obj)


## 3. Selecting Model and Simulating Tournament
### 3.1 Selecting Model

After comparing the test accuracies of all three models, including the parameter tuned Random Forest/ XGBoost models, the XGBoost model is proven to be the most accurate. With a 69.5% test accuracy score. While all the models favor choosing winners correctly than choosing losers correctly, the XGBoost model gives us the best of both worlds. We will start simulation with this model, than continue to run simulations with the other models just for fun.

### 3.2 Creating relevant simulation data



In [238]:
tsim = pd.read_csv('mmHistData/Tournament Simulation.csv')

tsim.head()

YEAR  BY YEAR NO  BY ROUND NO  TEAM NO         TEAM  SEED  ROUND  \
0  2025         NaN            0      NaN       Auburn     1      1   
1  2025         NaN            0      NaN  Alabama St.    16     64   
2  2025         NaN            0      NaN   Louisville     8      1   
3  2025         NaN            0      NaN    Creighton     9     64   
4  2025         NaN            0      NaN     Michigan     5      1   

   CURRENT ROUND  
0             64  
1             64  
2             64  
3             64  
4             64

In [239]:
tsim_df = tsim.drop(columns=['YEAR', 'BY YEAR NO', 'BY ROUND NO', 'TEAM NO'])
tsim_df.head()

TEAM  SEED  ROUND  CURRENT ROUND
0       Auburn     1      1             64
1  Alabama St.    16     64             64
2   Louisville     8      1             64
3    Creighton     9     64             64
4     Michigan     5      1             64

In [240]:
# Read in current Kenpom stats of all D1 teams
currKenPom = pd.read_csv('kenpom-ncaa-2025.csv')

currKenPom.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 347 entries, 0 to 346
Data columns (total 13 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Unnamed: 0    347 non-null    int64  
 1   Rk            347 non-null    float64
 2   Team          347 non-null    object 
 3   Conf          347 non-null    object 
 4   NetRtg        347 non-null    float64
 5   ORtg          347 non-null    float64
 6   DRtg          347 non-null    float64
 7   AdjT          347 non-null    float64
 8   Luck          347 non-null    float64
 9   SS_NetRtg     347 non-null    float64
 10  SS_ORtg       347 non-null    float64
 11  SS_DRtg       347 non-null    float64
 12  NCSOS_NetRtg  347 non-null    float64
dtypes: float64(10), int64(1), object(2)
memory usage: 35.4+ KB


In [241]:
#Drop irrelevant columns
currKenPom_df = currKenPom.drop(columns=['Rk','Conf', 'NetRtg', 'Luck', 'SS_NetRtg', 'SS_ORtg', 'SS_DRtg', 'NCSOS_NetRtg'])

currKenPom_df.head()

Unnamed: 0       Team   ORtg  DRtg  AdjT
0           0       Duke  129.1  89.7  66.3
1           1     Auburn  130.7  93.0  68.6
2           2    Houston  124.4  89.2  61.3
3           3    Florida  126.8  92.9  69.5
4           4  Tennessee  118.5  87.1  64.0

In [242]:
#Merge tsim_df with currKenPom_df
tournament_df = tsim_df.merge(currKenPom_df, left_on='TEAM', right_on='Team', how='left')

tournament_df.head()

TEAM  SEED  ROUND  CURRENT ROUND  Unnamed: 0         Team   ORtg  \
0       Auburn     1      1             64           1       Auburn  130.7   
1  Alabama St.    16     64             64         278  Alabama St.  102.3   
2   Louisville     8      1             64          23   Louisville  118.8   
3    Creighton     9     64             64          36    Creighton  116.1   
4     Michigan     5      1             64          28     Michigan  117.1   

    DRtg  AdjT  
0   93.0  68.6  
1  113.1  68.1  
2   96.3  69.2  
3   98.0  68.0  
4   95.9  69.9

In [243]:
#Clean Cols
tournament_df = tournament_df.drop(columns=['Team', 'Unnamed: 0'])

#drop rows with NA values, meaning not participating in the tournament
tournament_df = tournament_df.dropna()

print(tournament_df)


           TEAM  SEED  ROUND  CURRENT ROUND   ORtg   DRtg  AdjT
0        Auburn     1      1             64  130.7   93.0  68.6
1   Alabama St.    16     64             64  102.3  113.1  68.1
2    Louisville     8      1             64  118.8   96.3  69.2
3     Creighton     9     64             64  116.1   98.0  68.0
4      Michigan     5      1             64  117.1   95.9  69.9
..          ...   ...    ...            ...    ...    ...   ...
59         Troy    14     64             64  106.8  101.8  66.3
60         UCLA     7      1             64  116.8   95.1  65.0
61     Utah St.    10     64             64  121.0  106.3  68.2
62    Tennessee     2      1             64  118.5   87.1  64.0
63      Wofford    15     64             64  112.2  109.4  64.4

[64 rows x 7 columns]


In [244]:
#Must reshape the data to match format of training data
tournament_df = tournament_df.reset_index(drop=True)

matchups = []

for i in range(0, len(tournament_df), 2):  # Step by 2 to group consecutive pairs
    if i + 1 < len(tournament_df):  # Ensure we don't go out of bounds
        matchups.append({
            'Team_1': tournament_df.loc[i, 'TEAM'],
            'Seed_1': tournament_df.loc[i, 'SEED'],
            'Team_1_ORtg': tournament_df.loc[i, 'ORtg'],
            'Team_1_DRtg': tournament_df.loc[i, 'DRtg'],
            'Team_1_AdjT': tournament_df.loc[i, 'AdjT'],
            'Team_2': tournament_df.loc[i + 1, 'TEAM'],
            'Seed_2': tournament_df.loc[i + 1, 'SEED'],
            'Team_2_ORtg': tournament_df.loc[i + 1, 'ORtg'],
            'Team_2_DRtg': tournament_df.loc[i + 1, 'DRtg'],
            'Team_2_AdjT': tournament_df.loc[i + 1, 'AdjT']
        })

# Convert the list into a DataFrame
matchups_df = pd.DataFrame(matchups)
 
matchups_df.head()  





Team_1  Seed_1  Team_1_ORtg  Team_1_DRtg  Team_1_AdjT          Team_2  \
0       Auburn       1        130.7         93.0         68.6     Alabama St.   
1   Louisville       8        118.8         96.3         69.2       Creighton   
2     Michigan       5        117.1         95.9         69.9    UC San Diego   
3    Texas A&M       4        115.9         93.0         66.8            Yale   
4  Mississippi       6        118.0         96.0         68.4  North Carolina   

   Seed_2  Team_2_ORtg  Team_2_DRtg  Team_2_AdjT  
0      16        102.3        113.1         68.1  
1       9        116.1         98.0         68.0  
2      12        116.2         97.7         66.3  
3      13        115.9        104.7         68.1  
4      11        119.2        102.0         71.0

### 3.3 Simulating Tournament Structure

In [254]:
#Simulates entire bracket with XGBoost model
def simulate_tournament(matchups_df, model):
    round_num = 1

    #store stats in dictionary for quicker lookup
    team_stats_dict = {}
    for i, row in matchups_df.iterrows():
        team_stats_dict[row['Team_1']] = {
            'AdjO': row['Team_1_ORtg'],  # Store original stat, not the diff
            'AdjD': row['Team_1_DRtg'],
            'AdjT': row['Team_1_AdjT'],
            'Seed': row['Seed_1'],
        }
        team_stats_dict[row['Team_2']] = {
            'AdjO': row['Team_2_ORtg'],
            'AdjD': row['Team_2_DRtg'],
            'AdjT': row['Team_2_AdjT'],
            'Seed': row['Seed_2'],
        }
    
    while len(matchups_df) > 1:
        print(f"\n🏀 **Round {round_num} - {len(matchups_df)*2} Teams Remaining** 🏀\n")

        winners = []
        next_round_matchups = []

        for i,(index,row) in enumerate(matchups_df.iterrows()):
            # set teams
            team1 = row['Team_1']
            team2 = row['Team_2']

            #pull team stats from dictionary
            team1_stats = team_stats_dict[team1]
            team2_stats = team_stats_dict[team2] 

             # Extract the differential stats (adjOff, adjDef, adjT, seedDiff)
            adjO_diff = team1_stats['AdjO'] - team2_stats['AdjO']
            adjD_diff = team1_stats['AdjD'] - team2_stats['AdjD']
            adjT_diff = team1_stats['AdjT'] - team2_stats['AdjT']
            seed_diff = team1_stats['Seed'] - team2_stats['Seed']
            
            X_matchup = [[adjO_diff, adjD_diff, adjT_diff, seed_diff]]
            y_pred = model.predict(X_matchup)  

            if y_pred == 1:
                winner = team1
            else:
                winner = team2

            print(f"✅ {team1} vs {team2} → **{winner} WINS!**")
            winners.append(winner)
            
            #check for winner, else crate new matchups for next round
            if len(winners) % 2 == 0:
                next_round_matchups.append({
                    'Team_1': winners[-2],
                    'Team_2': winners[-1],
                })

        matchups_df = pd.DataFrame(next_round_matchups)
        round_num += 1
    
    final_team1 = matchups_df.iloc[0]['Team_1']
    final_team2 = matchups_df.iloc[0]['Team_2']
    print(f"\n🏀 **Final Matchup** 🏀\n✅ {final_team1} vs {final_team2} → **{final_team1 if y_pred == 1 else final_team2} WINS!**")

    champion = matchups_df.iloc[0]['Team_1']
    print(f"\n🏆 **{champion} WINS THE TOURNAMENT!** 🏆")
        

## 4. Testing Model on 2025 Bracket with XGBoost, Random Forest, & Linear Regression

### 4.1 XGBoost

In [255]:
simulate_tournament(matchups_df, best_xgb_model)


🏀 **Round 1 - 64 Teams Remaining** 🏀

✅ Auburn vs Alabama St. → **Auburn WINS!**
✅ Louisville vs Creighton → **Louisville WINS!**
✅ Michigan vs UC San Diego → **Michigan WINS!**
✅ Texas A&M vs Yale → **Texas A&M WINS!**
✅ Mississippi vs North Carolina → **Mississippi WINS!**
✅ Iowa St. vs Lipscomb → **Iowa St. WINS!**
✅ Marquette vs New Mexico → **Marquette WINS!**
✅ Michigan St. vs Bryant → **Michigan St. WINS!**
✅ Florida vs Norfolk St. → **Florida WINS!**
✅ Connecticut vs Oklahoma → **Connecticut WINS!**
✅ Memphis vs Colorado St. → **Memphis WINS!**
✅ Maryland vs Grand Canyon → **Maryland WINS!**
✅ Missouri vs Drake → **Missouri WINS!**
✅ Texas Tech vs UNC Wilmington → **Texas Tech WINS!**
✅ Kansas vs Arkansas → **Kansas WINS!**
✅ St. John's vs Nebraska Omaha → **St. John's WINS!**
✅ Duke vs American → **Duke WINS!**
✅ Mississippi St. vs Baylor → **Baylor WINS!**
✅ Oregon vs Liberty → **Oregon WINS!**
✅ Arizona vs Akron → **Arizona WINS!**
✅ BYU vs VCU → **BYU WINS!**
✅ Wisconsin v

### 4.2 Random Forest

In [257]:
simulate_tournament(matchups_df, best_rf_model)


🏀 **Round 1 - 64 Teams Remaining** 🏀

✅ Auburn vs Alabama St. → **Auburn WINS!**
✅ Louisville vs Creighton → **Louisville WINS!**
✅ Michigan vs UC San Diego → **Michigan WINS!**
✅ Texas A&M vs Yale → **Texas A&M WINS!**
✅ Mississippi vs North Carolina → **Mississippi WINS!**
✅ Iowa St. vs Lipscomb → **Iowa St. WINS!**
✅ Marquette vs New Mexico → **Marquette WINS!**
✅ Michigan St. vs Bryant → **Michigan St. WINS!**
✅ Florida vs Norfolk St. → **Florida WINS!**
✅ Connecticut vs Oklahoma → **Connecticut WINS!**
✅ Memphis vs Colorado St. → **Memphis WINS!**
✅ Maryland vs Grand Canyon → **Maryland WINS!**
✅ Missouri vs Drake → **Missouri WINS!**
✅ Texas Tech vs UNC Wilmington → **Texas Tech WINS!**
✅ Kansas vs Arkansas → **Kansas WINS!**
✅ St. John's vs Nebraska Omaha → **St. John's WINS!**
✅ Duke vs American → **Duke WINS!**
✅ Mississippi St. vs Baylor → **Baylor WINS!**
✅ Oregon vs Liberty → **Oregon WINS!**
✅ Arizona vs Akron → **Arizona WINS!**
✅ BYU vs VCU → **BYU WINS!**
✅ Wisconsin v

C:\Users\shash\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\LocalCache\local-packages\Python311\site-packages\sklearn\utils\validation.py:2739: UserWarning: X does not have valid feature names, but RandomForestClassifier was fitted with feature names
  warnings.warn(
C:\Users\shash\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\LocalCache\local-packages\Python311\site-packages\sklearn\utils\validation.py:2739: UserWarning: X does not have valid feature names, but RandomForestClassifier was fitted with feature names
  warnings.warn(
C:\Users\shash\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\LocalCache\local-packages\Python311\site-packages\sklearn\utils\validation.py:2739: UserWarning: X does not have valid feature names, but RandomForestClassifier was fitted with feature names
  warnings.warn(
C:\Users\shash\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\LocalCache\local-pac

### 4.3 Linear Regression

In [258]:
simulate_tournament(matchups_df, model)


🏀 **Round 1 - 64 Teams Remaining** 🏀

✅ Auburn vs Alabama St. → **Auburn WINS!**
✅ Louisville vs Creighton → **Louisville WINS!**
✅ Michigan vs UC San Diego → **Michigan WINS!**
✅ Texas A&M vs Yale → **Texas A&M WINS!**
✅ Mississippi vs North Carolina → **Mississippi WINS!**
✅ Iowa St. vs Lipscomb → **Iowa St. WINS!**
✅ Marquette vs New Mexico → **Marquette WINS!**
✅ Michigan St. vs Bryant → **Michigan St. WINS!**
✅ Florida vs Norfolk St. → **Florida WINS!**
✅ Connecticut vs Oklahoma → **Connecticut WINS!**
✅ Memphis vs Colorado St. → **Memphis WINS!**
✅ Maryland vs Grand Canyon → **Maryland WINS!**
✅ Missouri vs Drake → **Missouri WINS!**
✅ Texas Tech vs UNC Wilmington → **Texas Tech WINS!**
✅ Kansas vs Arkansas → **Kansas WINS!**
✅ St. John's vs Nebraska Omaha → **St. John's WINS!**
✅ Duke vs American → **Duke WINS!**
✅ Mississippi St. vs Baylor → **Baylor WINS!**
✅ Oregon vs Liberty → **Oregon WINS!**
✅ Arizona vs Akron → **Arizona WINS!**
✅ BYU vs VCU → **BYU WINS!**
✅ Wisconsin v

C:\Users\shash\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\LocalCache\local-packages\Python311\site-packages\sklearn\utils\validation.py:2739: UserWarning: X does not have valid feature names, but LogisticRegression was fitted with feature names
  warnings.warn(
C:\Users\shash\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\LocalCache\local-packages\Python311\site-packages\sklearn\utils\validation.py:2739: UserWarning: X does not have valid feature names, but LogisticRegression was fitted with feature names
  warnings.warn(
C:\Users\shash\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\LocalCache\local-packages\Python311\site-packages\sklearn\utils\validation.py:2739: UserWarning: X does not have valid feature names, but LogisticRegression was fitted with feature names
  warnings.warn(
C:\Users\shash\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\LocalCache\local-packages\Python

### 4.4 Reflection

All three different models have some differences in the preliminary rounds, but as the tournament narrows, we can see a heavy bias towards *seeding* in the elite 8 and final 4. To the point where in all models the final four are **all the 1 seeds.** I dont think this is an inheriently incorrect way to go about brackets, because historically using seeds is a good measure of how the tournament will move along. However I think there may be some benefits to training the models without seeding information, and see how the model predicts wins based on solely offensive, deffensive, and tempo efficiency from the KenPom statistics. There is a good chance taking seed out of account may make the overall bracket prediction worse, but it will still be fun to try out!!

## 5. Training models without seed differential information

### 5.1 Splitting Data

In [259]:
#Defining features(x) and target variable(y)
X = merged_df[['AdjO_Diff', 'AdjD_Diff', 'AdjT_Diff']]
y = merged_df['Winner']

#Splitting the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42, stratify=y)

# Print dataset sizes
print(f"Training size: {len(X_train)}, Testing size: {len(X_test)}")

Training size: 798, Testing size: 200


### 5.2 Logistic Regression

In [260]:
model = LogisticRegression()

model.fit(X_train, y_train)

train_accuracy = model.score(X_train, y_train)

print(f"Training Accuracy: {train_accuracy:.4f}")

# Make predictions on the test set
y_pred = model.predict(X_test)

# Calculate test accuracy
test_accuracy = accuracy_score(y_test, y_pred)
print(f"Test Accuracy: {test_accuracy:.4f}")

# Print classification report (detailed metrics)
print("\nClassification Report:\n", classification_report(y_test, y_pred))

# Print confusion matrix
print("\nConfusion Matrix:\n", confusion_matrix(y_test, y_pred))

Training Accuracy: 0.7155
Test Accuracy: 0.6800

Classification Report:
               precision    recall  f1-score   support

           0       0.61      0.45      0.52        76
           1       0.71      0.82      0.76       124

    accuracy                           0.68       200
   macro avg       0.66      0.63      0.64       200
weighted avg       0.67      0.68      0.67       200


Confusion Matrix:
 [[ 34  42]
 [ 22 102]]


### 5.3 Random Forest

In [261]:
# Initialize the Random Forest model
rf_model = RandomForestClassifier(random_state=42)

# Train the model on the training data
rf_model.fit(X_train, y_train)

# Print training accuracy
train_accuracy_rf = rf_model.score(X_train, y_train)
print(f"Training Accuracy (Random Forest): {train_accuracy_rf:.4f}")

# Make predictions on the test set
y_pred_rf = rf_model.predict(X_test)

# Calculate test accuracy
test_accuracy_rf = accuracy_score(y_test, y_pred_rf)
print(f"Test Accuracy (Random Forest): {test_accuracy_rf:.4f}")

# Print the classification report
print("\nClassification Report (Random Forest):\n", classification_report(y_test, y_pred_rf))

# Print confusion matrix
print("\nConfusion Matrix (Random Forest):\n", confusion_matrix(y_test, y_pred_rf))

# Define parameter grid for Random Forest
param_dist_rf = {
    'n_estimators': [50, 100, 200],
    'max_depth': [None, 10, 20, 30],
    'min_samples_split': [2, 5, 10],
    'min_samples_leaf': [1, 2, 4],
    'max_features': ['sqrt', 'log2']  # Valid options
}

# Initialize Random Forest model
rf_model = RandomForestClassifier(random_state=42)

# Initialize RandomizedSearchCV
random_search_rf = RandomizedSearchCV(estimator=rf_model, param_distributions=param_dist_rf, n_iter=100, cv=3, n_jobs=-1, verbose=2)

# Fit the model
random_search_rf.fit(X_train, y_train)

# Print the best parameters
print("Best parameters for Random Forest:", random_search_rf.best_params_)

# Get the best model
best_rf_model_no_seed = random_search_rf.best_estimator_

# Print test accuracy of the best model
test_accuracy_rf_tuned = best_rf_model_no_seed.score(X_test, y_test)
print(f"Tuned Random Forest Test Accuracy: {test_accuracy_rf_tuned:.4f}")




Training Accuracy (Random Forest): 1.0000
Test Accuracy (Random Forest): 0.6500

Classification Report (Random Forest):
               precision    recall  f1-score   support

           0       0.56      0.39      0.46        76
           1       0.68      0.81      0.74       124

    accuracy                           0.65       200
   macro avg       0.62      0.60      0.60       200
weighted avg       0.64      0.65      0.63       200


Confusion Matrix (Random Forest):
 [[ 30  46]
 [ 24 100]]
Fitting 3 folds for each of 100 candidates, totalling 300 fits
Best parameters for Random Forest: {'n_estimators': 50, 'min_samples_split': 10, 'min_samples_leaf': 4, 'max_features': 'log2', 'max_depth': 10}
Tuned Random Forest Test Accuracy: 0.6900


### 5.4 XGBoost

In [263]:
# Initialize XGBoost classifier
xgb_model = xgb.XGBClassifier(random_state=42, use_label_encoder=False, eval_metric='logloss')

# Train the model
xgb_model.fit(X_train, y_train)

# Calculate training accuracy
train_accuracy_xgb = xgb_model.score(X_train, y_train)
print(f"Training Accuracy (XGBoost): {train_accuracy_xgb:.4f}")

# Make predictions on the test set
y_pred_xgb = xgb_model.predict(X_test)

# Calculate test accuracy
test_accuracy_xgb = accuracy_score(y_test, y_pred_xgb)
print(f"Test Accuracy (XGBoost): {test_accuracy_xgb:.4f}")

# Print classification report
print("\nClassification Report (XGBoost):\n", classification_report(y_test, y_pred_xgb))

# Print confusion matrix
print("\nConfusion Matrix (XGBoost):\n", confusion_matrix(y_test, y_pred_xgb))

from sklearn.model_selection import RandomizedSearchCV
import xgboost as xgb

# Define parameter grid for XGBoost (using ranges)
param_dist_xgb = {
    'n_estimators': [50, 100, 200, 300],
    'max_depth': [3, 6, 9, 12],
    'learning_rate': [0.01, 0.1, 0.2],
    'subsample': [0.7, 0.8, 0.9, 1.0],
    'colsample_bytree': [0.7, 0.8, 0.9],
    'gamma': [0, 0.1, 0.2]
}

# Initialize XGBoost model
xgb_model = xgb.XGBClassifier(random_state=42, use_label_encoder=False, eval_metric='logloss')

# Initialize RandomizedSearchCV
random_search_xgb = RandomizedSearchCV(estimator=xgb_model, param_distributions=param_dist_xgb, n_iter=100, cv=3, n_jobs=-1, verbose=2)

# Fit the model
random_search_xgb.fit(X_train, y_train)

# Print best parameters
print("Best parameters for XGBoost:", random_search_xgb.best_params_)

# Get the best model
best_xgb_model_no_seed = random_search_xgb.best_estimator_

# Print test accuracy of the best model
test_accuracy_xgb_tuned = best_xgb_model_no_seed.score(X_test, y_test)
print(f"Tuned XGBoost Test Accuracy: {test_accuracy_xgb_tuned:.4f}")



C:\Users\shash\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\LocalCache\local-packages\Python311\site-packages\xgboost\training.py:183: UserWarning: [16:20:49] WARNING: C:\actions-runner\_work\xgboost\xgboost\src\learner.cc:738: 
Parameters: { "use_label_encoder" } are not used.

  bst.update(dtrain, iteration=i, fobj=obj)


Training Accuracy (XGBoost): 0.9962
Test Accuracy (XGBoost): 0.6650

Classification Report (XGBoost):
               precision    recall  f1-score   support

           0       0.58      0.41      0.48        76
           1       0.69      0.82      0.75       124

    accuracy                           0.67       200
   macro avg       0.64      0.62      0.62       200
weighted avg       0.65      0.67      0.65       200


Confusion Matrix (XGBoost):
 [[ 31  45]
 [ 22 102]]
Fitting 3 folds for each of 100 candidates, totalling 300 fits
Best parameters for XGBoost: {'subsample': 0.7, 'n_estimators': 50, 'max_depth': 3, 'learning_rate': 0.1, 'gamma': 0.2, 'colsample_bytree': 0.8}
Tuned XGBoost Test Accuracy: 0.6900


C:\Users\shash\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\LocalCache\local-packages\Python311\site-packages\xgboost\training.py:183: UserWarning: [16:20:52] WARNING: C:\actions-runner\_work\xgboost\xgboost\src\learner.cc:738: 
Parameters: { "use_label_encoder" } are not used.

  bst.update(dtrain, iteration=i, fobj=obj)


## 6. Simulating All models without Seed differential

### 6.1 Updating simulation code

In [265]:
#Simulates entire bracket with XGBoost model
def simulate_tournament(matchups_df, model):
    round_num = 1

    #store stats in dictionary for quicker lookup
    team_stats_dict = {}
    for i, row in matchups_df.iterrows():
        team_stats_dict[row['Team_1']] = {
            'AdjO': row['Team_1_ORtg'],  # Store original stat, not the diff
            'AdjD': row['Team_1_DRtg'],
            'AdjT': row['Team_1_AdjT'],
        }
        team_stats_dict[row['Team_2']] = {
            'AdjO': row['Team_2_ORtg'],
            'AdjD': row['Team_2_DRtg'],
            'AdjT': row['Team_2_AdjT'],
        }
    
    while len(matchups_df) > 1:
        print(f"\n🏀 **Round {round_num} - {len(matchups_df)*2} Teams Remaining** 🏀\n")

        winners = []
        next_round_matchups = []

        for i,(index,row) in enumerate(matchups_df.iterrows()):
            # set teams
            team1 = row['Team_1']
            team2 = row['Team_2']

            #pull team stats from dictionary
            team1_stats = team_stats_dict[team1]
            team2_stats = team_stats_dict[team2] 

             # Extract the differential stats (adjOff, adjDef, adjT)
            adjO_diff = team1_stats['AdjO'] - team2_stats['AdjO']
            adjD_diff = team1_stats['AdjD'] - team2_stats['AdjD']
            adjT_diff = team1_stats['AdjT'] - team2_stats['AdjT']
            
            X_matchup = [[adjO_diff, adjD_diff, adjT_diff]]
            y_pred = model.predict(X_matchup)  

            if y_pred == 1:
                winner = team1
            else:
                winner = team2

            print(f"✅ {team1} vs {team2} → **{winner} WINS!**")
            winners.append(winner)
            
            #check for winner, else crate new matchups for next round
            if len(winners) % 2 == 0:
                next_round_matchups.append({
                    'Team_1': winners[-2],
                    'Team_2': winners[-1],
                })

        matchups_df = pd.DataFrame(next_round_matchups)
        round_num += 1
    
    final_team1 = matchups_df.iloc[0]['Team_1']
    final_team2 = matchups_df.iloc[0]['Team_2']
    print(f"\n🏀 **Final Matchup** 🏀\n✅ {final_team1} vs {final_team2} → **{final_team1 if y_pred == 1 else final_team2} WINS!**")

    champion = matchups_df.iloc[0]['Team_1']
    print(f"\n🏆 **{champion} WINS THE TOURNAMENT!** 🏆")
        

### 6.2 Logistic Regression

In [266]:
simulate_tournament(matchups_df, model)


🏀 **Round 1 - 64 Teams Remaining** 🏀

✅ Auburn vs Alabama St. → **Auburn WINS!**
✅ Louisville vs Creighton → **Louisville WINS!**
✅ Michigan vs UC San Diego → **Michigan WINS!**
✅ Texas A&M vs Yale → **Texas A&M WINS!**
✅ Mississippi vs North Carolina → **Mississippi WINS!**
✅ Iowa St. vs Lipscomb → **Iowa St. WINS!**
✅ Marquette vs New Mexico → **Marquette WINS!**
✅ Michigan St. vs Bryant → **Michigan St. WINS!**
✅ Florida vs Norfolk St. → **Florida WINS!**
✅ Connecticut vs Oklahoma → **Connecticut WINS!**
✅ Memphis vs Colorado St. → **Memphis WINS!**
✅ Maryland vs Grand Canyon → **Maryland WINS!**
✅ Missouri vs Drake → **Missouri WINS!**
✅ Texas Tech vs UNC Wilmington → **Texas Tech WINS!**
✅ Kansas vs Arkansas → **Kansas WINS!**
✅ St. John's vs Nebraska Omaha → **St. John's WINS!**
✅ Duke vs American → **Duke WINS!**
✅ Mississippi St. vs Baylor → **Baylor WINS!**
✅ Oregon vs Liberty → **Oregon WINS!**
✅ Arizona vs Akron → **Arizona WINS!**
✅ BYU vs VCU → **BYU WINS!**
✅ Wisconsin v

C:\Users\shash\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\LocalCache\local-packages\Python311\site-packages\sklearn\utils\validation.py:2739: UserWarning: X does not have valid feature names, but LogisticRegression was fitted with feature names
  warnings.warn(
C:\Users\shash\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\LocalCache\local-packages\Python311\site-packages\sklearn\utils\validation.py:2739: UserWarning: X does not have valid feature names, but LogisticRegression was fitted with feature names
  warnings.warn(
C:\Users\shash\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\LocalCache\local-packages\Python311\site-packages\sklearn\utils\validation.py:2739: UserWarning: X does not have valid feature names, but LogisticRegression was fitted with feature names
  warnings.warn(
C:\Users\shash\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\LocalCache\local-packages\Python

### 6.3 Random Forest

In [267]:
simulate_tournament(matchups_df, best_rf_model_no_seed)


🏀 **Round 1 - 64 Teams Remaining** 🏀

✅ Auburn vs Alabama St. → **Auburn WINS!**
✅ Louisville vs Creighton → **Louisville WINS!**
✅ Michigan vs UC San Diego → **Michigan WINS!**
✅ Texas A&M vs Yale → **Texas A&M WINS!**
✅ Mississippi vs North Carolina → **Mississippi WINS!**
✅ Iowa St. vs Lipscomb → **Iowa St. WINS!**
✅ Marquette vs New Mexico → **Marquette WINS!**
✅ Michigan St. vs Bryant → **Michigan St. WINS!**
✅ Florida vs Norfolk St. → **Florida WINS!**
✅ Connecticut vs Oklahoma → **Connecticut WINS!**
✅ Memphis vs Colorado St. → **Memphis WINS!**
✅ Maryland vs Grand Canyon → **Maryland WINS!**
✅ Missouri vs Drake → **Missouri WINS!**
✅ Texas Tech vs UNC Wilmington → **Texas Tech WINS!**
✅ Kansas vs Arkansas → **Kansas WINS!**
✅ St. John's vs Nebraska Omaha → **St. John's WINS!**
✅ Duke vs American → **Duke WINS!**
✅ Mississippi St. vs Baylor → **Mississippi St. WINS!**
✅ Oregon vs Liberty → **Oregon WINS!**
✅ Arizona vs Akron → **Arizona WINS!**
✅ BYU vs VCU → **VCU WINS!**
✅ Wi

C:\Users\shash\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\LocalCache\local-packages\Python311\site-packages\sklearn\utils\validation.py:2739: UserWarning: X does not have valid feature names, but RandomForestClassifier was fitted with feature names
  warnings.warn(
C:\Users\shash\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\LocalCache\local-packages\Python311\site-packages\sklearn\utils\validation.py:2739: UserWarning: X does not have valid feature names, but RandomForestClassifier was fitted with feature names
  warnings.warn(
C:\Users\shash\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\LocalCache\local-packages\Python311\site-packages\sklearn\utils\validation.py:2739: UserWarning: X does not have valid feature names, but RandomForestClassifier was fitted with feature names
  warnings.warn(
C:\Users\shash\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\LocalCache\local-pac

### 6.4 XGBoost 

In [268]:
simulate_tournament(matchups_df, best_xgb_model_no_seed)


🏀 **Round 1 - 64 Teams Remaining** 🏀

✅ Auburn vs Alabama St. → **Auburn WINS!**
✅ Louisville vs Creighton → **Louisville WINS!**
✅ Michigan vs UC San Diego → **Michigan WINS!**
✅ Texas A&M vs Yale → **Texas A&M WINS!**
✅ Mississippi vs North Carolina → **Mississippi WINS!**
✅ Iowa St. vs Lipscomb → **Iowa St. WINS!**
✅ Marquette vs New Mexico → **Marquette WINS!**
✅ Michigan St. vs Bryant → **Michigan St. WINS!**
✅ Florida vs Norfolk St. → **Florida WINS!**
✅ Connecticut vs Oklahoma → **Connecticut WINS!**
✅ Memphis vs Colorado St. → **Memphis WINS!**
✅ Maryland vs Grand Canyon → **Maryland WINS!**
✅ Missouri vs Drake → **Missouri WINS!**
✅ Texas Tech vs UNC Wilmington → **Texas Tech WINS!**
✅ Kansas vs Arkansas → **Kansas WINS!**
✅ St. John's vs Nebraska Omaha → **St. John's WINS!**
✅ Duke vs American → **Duke WINS!**
✅ Mississippi St. vs Baylor → **Baylor WINS!**
✅ Oregon vs Liberty → **Oregon WINS!**
✅ Arizona vs Akron → **Arizona WINS!**
✅ BYU vs VCU → **BYU WINS!**
✅ Wisconsin v

## 7. Reflection

Taking a look at the predictions for the tournament without including any of the seeding differentials, we can see that there is significant difference between the XGBoost models and Random Forest models (the logistic regression interestingly stays the same), however despite these difference the #1 seeded teams still managed to all make it into the final four with the model extremely confident with Auburn being the ones to cut down the net at the end of the tournament this year. 

Reasons for the heavy favoring towards seeded teams even without seeding differentials can maybe be explained by the KenPom statistics exaggerating the efficiencies for the #1 seeded teams. **The last and only time all 1 seeds made the final four was 2008**, but maybe history is going to be made again this year. However I still believe that the model is not completely incorrect for favoring higher seeds, as more often than not this is the pattern to most matchups. Upsets are hard to predict, this is evident in the accuracy rating for the models aswell, where all models consistently had a hard time predicting when a team would lose compared to when they would win.

If I was to make predictive models again for march madness, I think I would incorporate some additional statistics on top of the Ken Pom stats. I think stats like Momentum(*record for latest 5 games*), Strength of Schedule, and number of star players would impact the model to be more diverse with its picks approaching the elite eight segment of the tournament. I will be entering all the saved brackets into a competition and we will see which model did the best.